crates/wasmtime/src/runtime/gc.rs

```diff
@@ -76,20 +77,49 @@ impl<T> fmt::Debug for GcHeapOutOfMemory<T> {
 
 impl<T> fmt::Display for GcHeapOutOfMemory<T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "GC heap out of memory")
+        write!(
+            f,
+            "GC heap out of memory: no capacity for allocation of {} bytes",
+            self.bytes_needed
+        )
     }
 }
 
 impl<T> core::error::Error for GcHeapOutOfMemory<T> {}
 
 impl<T> GcHeapOutOfMemory<T> {
     #[cfg(feature = "gc")]
-    pub(crate) fn new(inner: T) -> Self {
-        Self { inner }
+    pub(crate) fn new(inner: T, bytes_needed: u64) -> Self {
+        Self {
+            inner,
+            bytes_needed,
+        }
+    }
+
+    #[cfg(feature = "gc")]
+    pub(crate) fn bytes_needed(&self) -> u64 {
+        self.bytes_needed
     }
 
     /// Recover this error's inner host value.
     pub fn into_inner(self) -> T {
         self.inner
     }
+
+    /// Take this error's inner host value, but also retain this
+    /// `GcHeapOutOfMemory` with `T` replaced with `()`.
+    ///
+    /// This allows you to both extract the inner `T` if necessary, and also
+    /// pass the `GcHeapOutOfMemory` error to [`Store::gc`][crate::Store::gc]
+    /// calls.
+    pub fn take_inner(self) -> (T, GcHeapOutOfMemory<()>) {
+        #[cfg(feature = "gc")]
+        {
+            (self.inner, GcHeapOutOfMemory::new((), self.bytes_needed))
+        }
+        #[cfg(not(feature = "gc"))]
+        {
+            unreachable!()
+        }
```

Comment:
Can you add some comments here for why the `unreachable!()` isn't reachable? It wasn't immediately obvious to me until I saw that the `new` constructor is gated, so presumably it's not possible to call this method when `gc` is disabled, but I think a comment here might be good to have